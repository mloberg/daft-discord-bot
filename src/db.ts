import { PrismaClient } from '@prisma/client';

const client = new PrismaClient({
    // TODO: setup logging
});

export default client;
