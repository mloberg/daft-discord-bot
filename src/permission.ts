import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
