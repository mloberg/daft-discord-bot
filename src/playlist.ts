import fs from 'fs-extra';
import { difference } from 'lodash';

import config from './config';
import { Dictionary } from './types';

interface Song {
    title?: string;
    file: string;
    tags: string[];
}

export class Manager {
    private playing: Dictionary<string | null> = {};
    private playlists: Dictionary<string[]> = {};

    constructor(private readonly file: string) {}

    async addSong(file: string, tags: string[], title?: string): Promise<void> {
        const songs = await this.getSongs();
        songs.push({ title, file, tags });
        await this.saveSongs(songs);
    }

    async findSongs(tags: string[]): Promise<string[]> {
        const songs = await this.getSongs();

        return songs.filter((s) => difference(tags, s.tags).length === 0).map((s) => s.file);
    }

    async updateSong(file: string, tags: string[], title?: string): Promise<void> {
        const songs = await this.getSongs();
        const index = songs.findIndex((s) => file === s.file);
        if (-1 === index) {
            throw new Error(`Could not find song for "${file}"`);
        }

        songs[index].title = title;
        songs[index].tags = tags;

        await this.saveSongs(songs);
    }

    clear(guild: string, room: string): void {
        const key = `${guild}_${room}`;
        delete this.playlists[key];
        delete this.playing[key];
    }

    create(guild: string, room: string, songs: string[]): void {
        this.playlists[`${guild}_${room}`] = songs;
    }

    next(guild: string, room: string): string | null {
        const key = `${guild}_${room}`;
        const playlist = this.playlists[key];
        if (!playlist) {
            return null;
        }

        return (this.playing[key] = playlist.shift() || null);
    }

    async nowPlaying(guild: string, room: string): Promise<Song | null> {
        const playing = this.playing[`${guild}_${room}`];
        if (!playing) {
            return null;
        }

        return (await this.getSongs()).find((s) => playing === s.file) || null;
    }

    async getSongs(): Promise<Song[]> {
        return fs.existsSync(this.file) ? await fs.readJSON(this.file) : [];
    }

    private async saveSongs(songs: Song[]) {
        await fs.writeJSON(this.file, songs);
    }
}

export default new Manager(config.dataFile);
