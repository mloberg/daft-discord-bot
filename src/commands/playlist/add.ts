import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import player from '../../player';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist add',
    alias: ['playlist append'],
    description: 'Append songs to a playlist',
    usage: '<name> [...songs]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const [name, ...append] = args._;
        if (!name) {
            throw new FriendlyError('I need a name for the playlist.');
        }

        const unsupported = append.filter((s) => !player.supports(s));
        if (unsupported.length !== 0) {
            throw new FriendlyError(`Cannot add to playlist. These songs are unsupported: ${unsupported.join(', ')}`);
        }

        const playlist = await db.playlist.findFirst({ where: { name, guild } });
        if (!playlist) {
            throw new FriendlyError("I couldn't find that playlist.");
        }

        const songs: string[] = playlist.songs as string[];
        songs.push(...append);

        await db.playlist.update({
            where: { id: playlist.id },
            data: { songs },
        });
        logger.info(`Added songs to playlist ${name} in ${message.member.guild.name}`);

        return message.react('ðŸŽµ');
    },
};

export default command;
