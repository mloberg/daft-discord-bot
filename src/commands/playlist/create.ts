import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import player from '../../player';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist create',
    alias: ['playlist new'],
    description: 'Create a new playlist',
    usage: '<name> [...songs]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const [name, ...songs] = args._;
        if (!name) {
            throw new FriendlyError('I need a name for the playlist.');
        }

        const unsupported = songs.filter((s) => !player.supports(s));
        if (unsupported.length !== 0) {
            throw new FriendlyError(`Cannot create playlist. These songs are unsupported: ${unsupported.join(', ')}`);
        }

        const existing = await db.playlist.findFirst({ where: { name, guild } });
        if (existing) {
            throw new FriendlyError('There is already a playlist with that name.');
        }

        await db.playlist.create({ data: { name, guild, songs } });
        logger.info(`Created playlist ${name} in ${message.member.guild.name}`);

        return message.react('ðŸŽµ');
    },
};

export default command;
