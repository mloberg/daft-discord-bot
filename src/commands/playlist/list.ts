import db from '../../db';
import { FriendlyError } from '../../error';
import { hasPermission } from '../../permission';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist list',
    alias: ['playlist show'],
    description: 'List all playlists or songs in a playlist',
    usage: '[name]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const name = args._.shift();
        if (!name) {
            const playlists = await db.playlist.findMany({ where: { guild } });

            return message.reply(playlists.map((playlist) => playlist.name).join(', '));
        }

        const playlist = await db.playlist.findFirst({ where: { name, guild } });
        if (!playlist) {
            throw new FriendlyError("I couldn't find that playlist.");
        }

        const songs = playlist.songs as string[];
        return message.reply(songs.join(', '));
    },
};

export default command;
