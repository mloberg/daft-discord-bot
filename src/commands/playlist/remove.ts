import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist delete',
    alias: ['playlist remove'],
    description: 'Delete a playlist',
    usage: '<name>',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const name = args._.shift();
        if (!name) {
            throw new FriendlyError('Give me a playlist to delete.');
        }

        await db.playlist.deleteMany({ where: { name, guild } });
        logger.info(`Deleted playlist ${name} from ${message.member.guild.name}`);

        return message.react('ðŸŽµ');
    },
};

export default command;
