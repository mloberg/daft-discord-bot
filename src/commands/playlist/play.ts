import { shuffle } from 'lodash';

import db from '../../db';
import { FriendlyError } from '../../error';
import { hasPermission } from '../../permission';
import playlist from '../../playlist';
import { Command } from '../../types';
import next from '../next';

const command: Command = {
    name: 'playlist play',
    alias: ['playlist start'],
    description: 'Play a playlist',
    usage: '<name> [--shuffle|-r] [--volume|-v <volume>]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        if (!message.member.voice.channel) {
            throw new FriendlyError('You are not in a voice channel.');
        }

        const guild = message.member.guild.id;
        const room = message.member.voice.channel.name;

        const name = args._.shift();
        if (!name) {
            throw new FriendlyError('Give me a playlist to play.');
        }

        const toPlay = await db.playlist.findFirst({ where: { guild, name } });
        if (!toPlay) {
            throw new FriendlyError("I couldn't find that playlist.");
        }

        const songs = toPlay.songs as string[];
        const randomize = args.shuffle || args.r;

        playlist.clear(guild, room);
        playlist.create(guild, room, randomize ? shuffle(songs) : songs);

        return await next.run(message, args);
    },
};

export default command;
