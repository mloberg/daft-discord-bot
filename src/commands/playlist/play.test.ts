import { Client, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import db from '../../db';
import { FriendlyError } from '../../error';
import { hasPermission } from '../../permission';
import Player from '../../player';
import playlist from '../../playlist';
import Next from '../next';
import command from './play';

const mocks = {
    next: mocked(Next),
    permission: mocked(hasPermission),
    player: mocked(Player),
};

jest.mock('discord.js', () => ({
    Message: jest.fn().mockImplementation(() => ({
        member: {
            guild: {
                id: 'testing',
            },
            voice: {
                channel: {
                    name: 'daft-test',
                },
            },
        },
    })),
}));

jest.mock('../next');
jest.mock('../../permission');
jest.mock('../../player');

describe('_playlist play', () => {
    const client = {} as Client;
    const channel = {} as TextChannel;

    beforeEach(async () => {
        mocks.next.run.mockClear();
        mocks.permission.mockClear();
        mocks.player.supports.mockClear();

        await db.playlist.create({ data: { name: 'test', guild: 'testing', songs: ['foo.mp3', 'bar.mp3'] } });
        await db.playlist.create({ data: { name: 'awesome playlist', guild: 'foo', songs: ['test.mp3'] } });
    });

    afterEach(async () => {
        playlist.clear('testing', 'daft-test');
        await db.$executeRaw`DELETE FROM playlists`;
        await db.$disconnect();
    });

    it('is a command', () => {
        expect(command.name).toBe('playlist play');
        expect(command).toMatchSnapshot();
    });

    it('plays a playlist', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: ['test'], $0: 'playlist play' });

        const songs = playlist.queue('testing', 'daft-test');
        expect(songs).toEqual(['foo.mp3', 'bar.mp3']);

        expect(mocks.next.run).toHaveBeenCalledTimes(1);
        expect(mocks.next.run).toHaveBeenCalledWith(message, { _: [], $0: 'playlist play' });
    });

    it('shuffles a playlist', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: ['test'], $0: 'playlist start', shuffle: true });

        const songs = playlist.queue('testing', 'daft-test');
        expect(songs).toHaveLength(2);
        expect(songs).toContain('foo.mp3');
        expect(songs).toContain('bar.mp3');

        expect(mocks.next.run).toHaveBeenCalledTimes(1);
        expect(mocks.next.run).toHaveBeenCalledWith(message, { _: [], $0: 'playlist start', shuffle: true });
    });

    it('will throw an error if no playlist is given', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await expect(command.run(message, { _: [], $0: 'playlist play' })).rejects.toThrow(
            new FriendlyError('Give me a playlist to play.'),
        );
    });

    it('will throw an error if no playlist was found', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await expect(command.run(message, { _: ['none'], $0: 'playlist play' })).rejects.toThrow(
            new FriendlyError("I couldn't find that playlist."),
        );
    });

    it('will not play a playlist from another guild', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await expect(command.run(message, { _: ['awesome playlist'], $0: 'playlist play' })).rejects.toThrow(
            new FriendlyError("I couldn't find that playlist."),
        );
    });

    it('will throw an error if not in a voice channel', async () => {
        (mocked(Message) as jest.Mock).mockImplementationOnce(() => {
            return { member: { voice: { channe: null } } };
        });
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await expect(command.run(message, { _: [], $0: 'playlist play' })).rejects.toThrow(
            new FriendlyError('You are not in a voice channel.'),
        );
    });

    it('will throw an error if does not have role', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(false);

        await expect(command.run(message, { _: [], $0: 'playlist play' })).rejects.toThrow(
            new FriendlyError('You do not have permission to do that.'),
        );
    });
});
