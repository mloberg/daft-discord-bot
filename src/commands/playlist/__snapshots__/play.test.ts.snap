// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_playlist play is a command 1`] = `
Object {
  "alias": Array [
    "playlist start",
  ],
  "description": "Play a playlist",
  "name": "playlist play",
  "run": [Function],
  "usage": "<name> [--shuffle|-r] [--volume|-v <volume>]",
}
`;
